### PR TITLE
Merged parts of spawnMicrobe and spawnBacteria

### LINKED ISSUE #625: Reduce bacteria related code duplication
Right now there are a few duplicated pieces of code for the bacteria spawning and the species updating. These should be merged to reduce clutter.

### PR DESCRIPTION
**Pull request by 1n48yg**
fixes #625

### CODE REVIEW

**Comment by hhyyrylainen** [comment]
I fixed the target branch (as it was already merged), master needs to be merged into this and the issues fixed
```diff
--- a/scripts/microbe_stage/microbe_operations.as
+++ b/scripts/microbe_stage/microbe_operations.as
@@ -914,7 +914,7 @@ ObjectID createMicrobe(CellStageWorld@ world, Float3 pos, const string &in speci
 
         return NULL_OBJECT;
     }
-    
+
     return _createMicrobeEntity(world, aiControlled, speciesName, false);
 }
 
@@ -934,7 +934,7 @@ ObjectID spawnMicrobe(CellStageWorld@ world, Float3 pos, const string &in specie
 
         return NULL_OBJECT;
     }
-    
+
     return _createMicrobeEntity(world, aiControlled, speciesName,
         // in_editor
         false);
@@ -947,7 +947,7 @@ ObjectID spawnMicrobe(CellStageWorld@ world, Float3 pos, const string &in specie
 {
     // Create microbeEntity with correct template, physics and species name
     auto microbeEntity = createMicrobe(world, pos, speciesName, aiControlled, "Microbe");
-    
+
     if (microbeEntity == NULL_OBJECT)
     {
         return NULL_OBJECT;
@@ -988,7 +988,7 @@ ObjectID spawnBacteria(CellStageWorld@ world, Float3 pos, const string &in speci
 {
     // Create microbeEntity with correct template, physics and species name
     auto microbeEntity = createMicrobe(world, pos, speciesName, aiControlled, "Bacteria");
-    
+
     if (microbeEntity == NULL_OBJECT)
     {
         return NULL_OBJECT;
```

**Comment by Untrustedlife** [comment]
> I fixed the target branch (as it was already merged), master needs to be merged into this and the issues fixed
> 
> ```diff
> --- a/scripts/microbe_stage/microbe_operations.as
> +++ b/scripts/microbe_stage/microbe_operations.as
> @@ -914,7 +914,7 @@ ObjectID createMicrobe(CellStageWorld@ world, Float3 pos, const string &in speci
>  
>          return NULL_OBJECT;
>      }
> -    
> +
>      return _createMicrobeEntity(world, aiControlled, speciesName, false);
>  }
>  
> @@ -934,7 +934,7 @@ ObjectID spawnMicrobe(CellStageWorld@ world, Float3 pos, const string &in specie
>  
>          return NULL_OBJECT;
>      }
> -    
> +
>      return _createMicrobeEntity(world, aiControlled, speciesName,
>          // in_editor
>          false);
> @@ -947,7 +947,7 @@ ObjectID spawnMicrobe(CellStageWorld@ world, Float3 pos, const string &in specie
>  {
>      // Create microbeEntity with correct template, physics and species name
>      auto microbeEntity = createMicrobe(world, pos, speciesName, aiControlled, "Microbe");
> -    
> +
>      if (microbeEntity == NULL_OBJECT)
>      {
>          return NULL_OBJECT;
> @@ -988,7 +988,7 @@ ObjectID spawnBacteria(CellStageWorld@ world, Float3 pos, const string &in speci
>  {
>      // Create microbeEntity with correct template, physics and species name
>      auto microbeEntity = createMicrobe(world, pos, speciesName, aiControlled, "Bacteria");
> -    
> +
>      if (microbeEntity == NULL_OBJECT)
>      {
>          return NULL_OBJECT;
> ```

Also they need to get rid of their processor component checking since that doesnt exist anymore at the species level.

    auto processor = getProcessorComponent(world, speciesName);

    if(processor is null){
        LOG_ERROR("Skipping " + typeOfMicrobe + " spawn because species '" + speciesName +
            "' doesn't have a processor component");

        return NULL_OBJECT;
    }

^ needs to go away

**Comment by 1n48yg** [comment]
> In that case maybe the method should be renamed spawnProkaryote hmmm

Ok sorry I phrased that wrong. I'm not saying we should change the entire system's name to prokaryote, I'm just saying the comments in that one specific area would benefit from being explained as prokaryotes. Because all it does is check if a cell has a nucleus. and if they don't spawn them at a decreased size, that description in my opinion fits prokaryote more than bacteria.

File: scripts/microbe_stage/microbe_operations.as [hunk 1]
```diff
@@ -895,22 +895,63 @@ bool validPlacement(CellStageWorld@ world, ObjectID microbeEntity, const Organel
...
+    }
+    
+    return _createMicrobeEntity(world, aiControlled, speciesName,
         // in_editor
         false);
+}
+
+// speciesName decides the template to use, while individualName is
+// used for referencing the instance
+ObjectID spawnMicrobe(CellStageWorld@ world, Float3 pos, const string &in speciesName,
```
Comment (hhyyrylainen) [inline scripts/microbe_stage/microbe_operations.as]: I think that this should be the *only* microbe spawning function, and it should automatically based on the species determine if this is a bacteria or not.

Comment (1n48yg) [reply scripts/microbe_stage/microbe_operations.as]: Alright, thanks for the feedback guys, you guys type really fast. I'll make the changes.

File: scripts/microbe_stage/microbe_operations.as [hunk 2]
```diff
@@ -895,22 +895,63 @@ bool validPlacement(CellStageWorld@ world, ObjectID microbeEntity, const Organel
...
+        LOG_WARNING("spawn" + typeOfMicrobe + ": spawning at y-coordinate: " + pos.Y);
+
+    auto processor = getProcessorComponent(world, speciesName);
 
+    if(processor is null){
+        LOG_ERROR("Skipping " + typeOfMicrobe + " spawn because species '" + speciesName +
+            "' doesn't have a processor component");
 
+        return NULL_OBJECT;
+    }
```
Comment (Untrustedlife) [inline scripts/microbe_stage/microbe_operations.as]: Species no longer has a processor component so this needs to be removed

File: scripts/microbe_stage/microbe_operations.as [hunk 3]
```diff
@@ -895,22 +895,63 @@ bool validPlacement(CellStageWorld@ world, ObjectID microbeEntity, const Organel
...
     auto microbeEntity = _createMicrobeEntity(world, aiControlled, speciesName,
+
+    auto processor = getProcessorComponent(world, speciesName);
+
+    if(processor is null){
+        LOG_ERROR("Skipping " + typeOfMicrobe + " spawn because species '" + speciesName +
+            "' doesn't have a processor component");
+
+        return NULL_OBJECT;
+    }
```
Comment (Untrustedlife) [inline scripts/microbe_stage/microbe_operations.as]: And also this

File: scripts/microbe_stage/microbe_operations.as (hunk 2 above)
Comment (Untrustedlife) [reply scripts/microbe_stage/microbe_operations.as]: (The whole check, and the grabbing of the processor before it)

File: scripts/microbe_stage/microbe_operations.as (hunk 3 above)
Comment (Untrustedlife) [reply scripts/microbe_stage/microbe_operations.as]: (The whole check, and the grabbing of the processor before it)

**Comment by Untrustedlife** [comment]
One other comment! More of a suggestion. In species_system.as mutateBacteria() and mutateSpecies are seperate methods (species mutation calls mutateBacteria after checking, instead of its own code,  but their code is completely identical aside from small things like a different id name for the species and bacteria and the fact that they call mutateMicrobe with true and false respectively, if you want to those two can be merged, if you dont want to do it however after you make my review changes i might do it and push to here. Or my own branch. 

  // Creates a mutated version of the species
    Species(SpeciesComponent@ parent, CellStageWorld@ world, bool isBacteria)
    {
        this.isBacteria = parent.isBacteria;
        LOG_INFO("Gene Code Is:"+parent.stringCode);
        if (!isBacteria)
        {
            name = randomSpeciesName();

        //Mutate the epithet
        if (GetEngine().GetRandom().GetNumber(0, 100) < MUTATION_WORD_EDIT){
            epithet = mutateWord(parent.epithet);
        }
        else {
            epithet = generateNameSection();
        }
        
            genus = parent.genus;
            colour=parent.colour;
            mutateBehavior(parent);

            // Make sure not over or under our scales
            cleanPersonality();


            // Subtly mutate color
            this.colour = Float4(parent.colour.X + randomMutationColourChannel(),
                parent.colour.Y + randomMutationColourChannel(),
                parent.colour.Z + randomMutationColourChannel(),
                parent.colour.W + randomMutationColourChannel());

            LOG_INFO("X:"+parent.colour.X+" Y:"+parent.colour.Y+" Z:"+parent.colour.Z+" W:"+parent.colour.W);
            LOG_INFO("X:"+colour.X+" Y:"+colour.Y+" Z:"+colour.Z+" W:"+colour.W);
            // Chance of new color needs to be low
            if (GetEngine().GetRandom().GetNumber(0,100) <= MUTATION_CHANGE_GENUS)
            {
                LOG_INFO("New Genus");
                // We can do more fun stuff here later
                if (GetEngine().GetRandom().GetNumber(0, 100) < MUTATION_WORD_EDIT){
                    genus = mutateWord(parent.genus);
                }
                else {
                    genus = generateNameSection();
                }
                // New genuses get to double their color change
                this.colour = Float4(parent.colour.X + randomMutationColourChannel(),
                    parent.colour.Y + randomMutationColourChannel(),
                    parent.colour.Z + randomMutationColourChannel(),
                    parent.colour.W + randomMutationColourChannel());
            }

            this.stringCode = mutateMicrobe(parent.stringCode,false);


            generateMembranes(parent);


            commonConstructor(world);


            this.setupSpawn(world);
        }
        else
        {
            mutateBacteria(parent,world);
        }
    }
    void mutateBacteria(SpeciesComponent@ parent, CellStageWorld@ world)
    {
        name = randomBacteriaName();
        genus = parent.genus;
        colour=parent.colour;
        //Mutate the epithet
        if (GetEngine().GetRandom().GetNumber(0, 100) < MUTATION_WORD_EDIT){
            epithet = mutateWord(parent.epithet);
        }
        else {
            epithet = generateNameSection();
        }


        mutateBehavior(parent);

        // Make sure not over or under our scales
        cleanPersonality();

        // Subtly mutate color
        this.colour = Float4(parent.colour.X + randomMutationColourChannel(),
            parent.colour.Y + randomMutationColourChannel(),
            parent.colour.Z + randomMutationColourChannel(),
            parent.colour.W + randomMutationColourChannel());


        LOG_INFO("X:"+parent.colour.X+" Y:"+parent.colour.Y+" Z:"+parent.colour.Z+" W:"+parent.colour.W);
        LOG_INFO("X:"+colour.X+" Y:"+colour.Y+" Z:"+colour.Z+" W:"+colour.W);

        if (GetEngine().GetRandom().GetNumber(0,100) <= MUTATION_CHANGE_GENUS)
        {
            LOG_INFO("New Genus of bacteria");

            // We can do more fun stuff here later
            //Mutate the genus
            if (GetEngine().GetRandom().GetNumber(0, 100) < MUTATION_WORD_EDIT){
                genus = mutateWord(parent.genus);
            }
            else {
                genus = generateNameSection();
            }

            // New genuses get to double color change
            this.colour = Float4(parent.colour.X + randomMutationColourChannel(),
                parent.colour.Y + randomMutationColourChannel(),
                parent.colour.Z + randomMutationColourChannel(),
                parent.colour.W + randomMutationColourChannel());
        }

        this.stringCode = mutateMicrobe(parent.stringCode,true);

        generateMembranes(parent);

        commonConstructor(world);
        
        this.setupSpawn(world);
    }

File: scripts/microbe_stage/microbe_operations.as [hunk 4]
```diff
@@ -895,29 +895,31 @@ bool validPlacement(CellStageWorld@ world, ObjectID microbeEntity, const Organel
...
 
     if(pos.Y != 0)
         LOG_WARNING("spawnMicrobe: spawning at y-coordinate: " + pos.Y);
 
-
     // Create microbeEntity with correct template, physics and species name
-    auto microbeEntity = _createMicrobeEntity(world, aiControlled, speciesName,
-        // in_editor
-        false);
+    auto microbeEntity = _createMicrobeEntity(world, aiControlled, speciesName, false);
```
Comment (hhyyrylainen) [inline scripts/microbe_stage/microbe_operations.as]: The comment should be preserved here. Anyway I don't think the parameter is used anymore, but as long as it exists, the comment should be kept.

File: scripts/microbe_stage/microbe_operations.as [hunk 5]
```diff
@@ -895,29 +895,31 @@ bool validPlacement(CellStageWorld@ world, ObjectID microbeEntity, const Organel
...
     microbePos._Position = pos;
     microbePos.Marked = true;
 
     auto physics = world.GetComponent_Physics(microbeEntity);
+    
+    auto speciesEntity = findSpeciesEntityByName(world, speciesName);
+    auto species = world.GetComponent_SpeciesComponent(speciesEntity);
+
+    if (species.isBacteria)
+    { physics.Body.SetMass(physics.Body.Mass * 10); }
```
Comment (hhyyrylainen) [inline scripts/microbe_stage/microbe_operations.as]: This formatting isn't what the C++ code would have. Also why does the bacteria get the mass multiplied by a ton?

File: scripts/microbe_stage/setup.as [hunk 6]
```diff
@@ -130,7 +130,7 @@ void setupPlayer(CellStageWorld@ world)
     GetThriveGame().playerData().lockedMap().addLock("chloroplast");
 
     ObjectID microbe = MicrobeOperations::spawnMicrobe(world, Float3(0, 0, 0), "Default",
-        false);
+        false, false);
```
Comment (hhyyrylainen) [inline scripts/microbe_stage/setup.as]: Having a default value for partOfColony would be nice to not have these changes.

File: scripts/microbe_stage/species_system.as [hunk 7]
```diff
@@ -337,66 +337,72 @@ class Species{
...
-             }
-            genus = parent.genus;
-            colour=parent.colour;
-            mutateBehavior(parent);
+        //Mutate the epithet
+        if (GetEngine().GetRandom().GetNumber(0, 100) < MUTATION_WORD_EDIT){
+            epithet = mutateWord(parent.epithet);
+        }
+        else {
+        epithet = generateNameSection();
```
Comment (hhyyrylainen) [inline scripts/microbe_stage/species_system.as]: Indentation

File: scripts/microbe_stage/microbe_operations.as (hunk 5 above)
Comment (1n48yg) [reply scripts/microbe_stage/microbe_operations.as]: I actually don't know why the bacteria's mass is multiplied by that much. This was the code in the previous spawnBacteria function, I just copied it over.

Comment (hhyyrylainen) [reply scripts/microbe_stage/microbe_operations.as]: That might be the change @Untrustedlife made to make the bacteria harder to push around. But anyway the formatting needs to be changed here.

File: scripts/microbe_stage/setup.as (hunk 6 above)
Comment (1n48yg) [reply scripts/microbe_stage/setup.as]: Just another question, I don't really understand what you mean by default value here. Should I override the function with a version that doesn't take this parameter? Or do you not need to enter all the parameters in angelscript?

Comment (hhyyrylainen) [reply scripts/microbe_stage/setup.as]: I'm pretty sure you can set default parameter values the same way in AS as in C++.

Comment (1n48yg) [reply scripts/microbe_stage/setup.as]: Alright thanks! I've made the requested changes, this should be good now.

**Comment by Untrustedlife** [comment]
There is one minor issue, player seems to move much slower now.

**Comment by Untrustedlife** [comment]
the issue was that set mass bit.

**Comment by Untrustedlife** [comment]
After my changes this is good. Ill go ahead and merge

**Pull request merged**